File: mvd/mvd/page/vbz_beratung/vbz_beratung.py
# -*- coding: utf-8 -*-
# Copyright (c) 2021, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from PyPDF2 import PdfFileWriter
from frappe.utils.data import add_days, getdate, now, today, now_datetime
from frappe.utils.pdf import get_file_data_from_writer

@frappe.whitelist()
def get_open_data():
    open_data = {
        'beratung': {
            'datenstand': now_datetime().strftime("%d.%m.%Y %H:%M:%S"),
            's': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Eingang', 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's1': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Eingang', 'mv_mitgliedschaft': ['is', 'not set'], 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's2': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Eingang', 'beratungskategorie': ['in', ['202 - MZ-Erhöhung', '300 - Nebenkosten']], 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's3': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Rückfrage: Termin vereinbaren', 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's4': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Rückfragen', 'kontaktperson': ['is', 'not set'], 'ungelesen': 0, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's5': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Rückfragen', 'kontaktperson': ['is', 'not set'], 'ungelesen': 1, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's6': len(frappe.get_list('Beratung', fields='name', filters={'status': ['not in', ['Rückfragen', 'Rückfrage: Termin vereinbaren', 'Eingang', 'Open', 'Zusammengeführt']], 'ungelesen': 1, 'kontaktperson': ['is', 'not set'], 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's7': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Open', 'kontaktperson': ['is', 'not set'], 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            's8': len(frappe.get_list('Beratung', fields='name', filters={'status': ['!=', 'Closed'], 's8': 1, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True, ignore_ifnull=True)),
            's9': len(frappe.get_list('Beratung', fields='name', filters={'ungelesen': 1, 'status': 'Zusammengeführt', 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True, ignore_ifnull=True)),
            'r': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Open', 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r1': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Open', 'beratung_prio': 'Hoch', 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r2': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Open', 'beratung_prio': ['!=', 'Hoch'], 'kontaktperson': 'Rechtsberatung Pool (MVBE)', 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r3': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Open', 'kontaktperson': ['!=', 'Rechtsberatung Pool (MVBE)'], 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r4': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Rückfragen', 'kontaktperson': ['is', 'set'], 'ungelesen': 0, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r5': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Rückfragen', 'kontaktperson': ['is', 'set'], 'ungelesen': 1, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r6': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Rückfragen', 'kontaktperson': ['!=', 'Rechtsberatung Pool (MVBE)'], 'kontaktperson': ['is', 'set'], 'ungelesen': 1, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r7': len(frappe.get_list('Beratung', fields='name', filters={'status': ['!=', 'Closed'], 'hat_termine': 1, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r8': len(frappe.get_list('Beratung', fields='name', filters={'status': 'Closed', 'hat_termine': 1, 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'r9': len(frappe.get_list('Beratung', fields='name', filters={'status': ['not in', ['Rückfragen', 'Open', 'Zusammengeführt']], 'ungelesen': 1, 'kontaktperson': ['is', 'set'], 'sektion_id': ['!=', 'MVDF']}, limit=100, distinct=True)),
            'p1': get_p1(frappe.session.user),
            'p2': get_p2(frappe.session.user),
            'p3': get_p3(frappe.session.user),
            'p4': get_p4(frappe.session.user)
        }
    }
    
    return open_data

def get_p1(user):
    p1_qty = frappe.db.sql("""SELECT COUNT(`name`) AS `qty` FROM `tabBeratung`
                                            WHERE `kontaktperson` IN (
                                                SELECT `parent`
                                                FROM `tabTermin Kontaktperson Multi User`
                                                WHERE `user` = '{user}'  
                                                AND `parent` NOT LIKE 'Rechtsberatung Pool (%)'
                                            )
                                            AND `status` = 'Open'
                                            AND `sektion_id` != 'MVDF'""".format(user=user), as_dict=True)[0].qty
    return p1_qty or 0

def get_p2(user):
    p2_qty = frappe.db.sql("""SELECT COUNT(`name`) AS `qty` FROM `tabBeratung`
                                            WHERE `kontaktperson` IN (
                                                SELECT `parent`
                                                FROM `tabTermin Kontaktperson Multi User`
                                                WHERE `user` = '{user}'  
                                                AND `parent` NOT LIKE 'Rechtsberatung Pool (%)'
                                            )
                                            AND `status` = 'Rückfragen'
                                            AND `ungelesen` = 0
                                            AND `sektion_id` != 'MVDF'""".format(user=user), as_dict=True)[0].qty
    return p2_qty or 0

def get_p3(user):
    p3_qty = frappe.db.sql("""SELECT COUNT(`name`) AS `qty` FROM `tabBeratung`
                                            WHERE `kontaktperson` IN (
                                                SELECT `parent`
                                                FROM `tabTermin Kontaktperson Multi User`
                                                WHERE `user` = '{user}'  AND `parent` NOT LIKE 'Rechtsberatung Pool (%)'
                                            )
                                            AND `status` = 'Rückfragen'
                                            AND `ungelesen` = 1
                                            AND `sektion_id` != 'MVDF'""".format(user=user), as_dict=True)[0].qty
    return p3_qty or 0

def get_p4(user):
    p4_qty = frappe.db.sql("""SELECT COUNT(`name`) AS `qty` FROM `tabBeratung`
                                            WHERE `kontaktperson` IN (
                                                SELECT `parent`
                                                FROM `tabTermin Kontaktperson Multi User`
                                                WHERE `user` = '{user}'  
                                                AND `parent` NOT LIKE 'Rechtsberatung Pool (%)'
                                            )
                                            AND `status` = 'Termin vereinbart'
                                            AND `hat_termine` = 1
                                            AND `sektion_id` != 'MVDF'""".format(user=user), as_dict=True)[0].qty
    return p4_qty or 0

@frappe.whitelist()
def get_user_kontaktperson(only_session_user=False):
    user_kontaktperson = frappe.db.sql("""SELECT `parent`
                                        FROM `tabTermin Kontaktperson Multi User`
                                        WHERE `user` = '{user}' 
                                        AND `parent` NOT LIKE 'Rechtsberatung Pool (%)'""".format(user=frappe.session.user), as_dict=True)
    user_kontaktpersonen = []
    for uk in user_kontaktperson:
        if not only_session_user:
            user_kontaktpersonen.append(uk.parent)
        else:
            if frappe.db.count('Termin Kontaktperson Multi User', {'parent': uk.parent}) < 2:
                user_kontaktpersonen.append(uk.parent)
    return user_kontaktpersonen
                

File: mvd/mvd/page/vbz_beratung/vbz_beratung.js
frappe.pages['vbz-beratung'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Verarbeitungszentrale',
        single_column: true
    });
    frappe.vbz_beratung.add_views(page);
    localStorage['firstLoad'] = true;
}
frappe.pages['vbz-beratung'].refresh= function(wrapper){
    frappe.vbz_beratung.show_view('vbz_beratung');
    frappe.dom.unfreeze();
} 

frappe.vbz_beratung = {
    add_views: function(page) {
        frappe.call({
            'method': "mvd.mvd.page.vbz_beratung.vbz_beratung.get_open_data",
            'args': {},
            'freeze': true,
            'freeze_message': 'Lade Verarbeitungszentrale...',
            'async': false,
            'callback': function(r)
            {
                if (r.message) {
                    page.add_view('vbz_beratung', frappe.render_template("vbz_beratung", eval(r.message.beratung)))
                    frappe.vbz_beratung.add_click_handlers(eval(r.message));
                    localStorage['firstLoad'] = true;
                }
            }
        });
    },
    show_view: function(view) {
        cur_page.page.page.set_view(view);
    },
    remove_click_handlers: function() {
        //
    },
    add_click_handlers: function(open_datas) {
        //~ frappe.vbz_beratung.remove_click_handlers();

        var sektion = frappe.boot.default_sektion || "MVBE";
        
        $("#s").click(function(){
            frappe.route_options = {"status": 'Eingang', "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s1").click(function(){
            frappe.route_options = {"status": 'Eingang', "mv_mitgliedschaft": ['is', 'not set'], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s2").click(function(){
            frappe.route_options = {"status": 'Eingang', "beratungskategorie": ['in', ['202 - MZ-Erhöhung', '300 - Nebenkosten', '202 - Mietzinserhöhung', '203 - Mietzinssenkung']], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s3").click(function(){
            frappe.route_options = {"status": 'Rückfrage: Termin vereinbaren', "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s4").click(function(){
            frappe.route_options = {"status": "Rückfragen", "kontaktperson": ['is', 'not set'], "ungelesen": 0, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s5").click(function(){
            frappe.route_options = {"status": "Rückfragen", "kontaktperson": ['is', 'not set'], "ungelesen": 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s6").click(function(){
            frappe.route_options = {"status": ["not in", ["Rückfragen", "Rückfrage: Termin vereinbaren", "Eingang", "Open", "Zusammengeführt"]], "ungelesen": 1, "kontaktperson": ['is', 'not set'], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s7").click(function(){
            frappe.route_options = {"status": ['in', ['Open', 'In Arbeit']], "kontaktperson": ['is', 'not set'], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s8").click(function(){
            frappe.route_options = {"status": ['!=', 'Closed'], "s8": 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s9").click(function(){
            frappe.route_options = {"status": 'Zusammengeführt', "ungelesen": 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#s10").click(function(){
            frappe.route_options = {"hat_termine": 1, "ungelesen": 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        
        $("#r").click(function(){
            frappe.route_options = {"status": ['in', ['Open', 'In Arbeit']], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r1").click(function(){
            frappe.route_options = {"status": ['in', ['Open', 'In Arbeit']], 'beratung_prio': 'Hoch', "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r2").click(function(){
             frappe.route_options = {'status': ['in', ['Open', 'In Arbeit']], 'kontaktperson': ['like', 'Rechtsberatung Pool%'], 'beratung_prio': ['!=', 'Hoch'], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r3").click(function(){
            frappe.route_options = {'status': ['in', ['Open', 'In Arbeit']], 'r3': 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r4").click(function(){
            frappe.route_options = {'status': 'Rückfragen', 'kontaktperson': ['is', 'set'], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r5").click(function(){
            frappe.route_options = {'status': 'Rückfragen', 'kontaktperson': ['is', 'set'], 'ungelesen': 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r6").click(function(){
            frappe.route_options = {'status': 'Rückfragen', 'kontaktperson': ['not like', 'Rechtsberatung Pool%'], 'kontaktperson': ['is', 'set'], 'ungelesen': 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r7").click(function(){
            frappe.route_options = {'status': ['!=', 'Closed'], 'hat_termine': 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r8").click(function(){
            frappe.route_options = {'status': 'Closed', 'hat_termine': 1, "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        $("#r9").click(function(){
            frappe.route_options = {"status": ["not in", ["Rückfragen", "Open", "Zusammengeführt", "Termin vereinbart"]], "ungelesen": 1, "kontaktperson": ['is', 'set'], "sektion_id": sektion}
            frappe.set_route("List", "Beratung", "List");
        });
        
        $("#p1").click(function(){
            frappe.call({
                'method': "mvd.mvd.page.vbz_beratung.vbz_beratung.get_user_kontaktperson",
                'args': {'only_session_user': 1},
                'async': false,
                'callback': function(r)
                {
                    frappe.route_options = {'status': ['in', ['Open', 'In Arbeit']], 'kontaktperson': ['in', r.message], "sektion_id": sektion}
                    frappe.set_route("List", "Beratung", "List");
                }
            });
        });
        $("#p2").click(function(){
            frappe.call({
                'method': "mvd.mvd.page.vbz_beratung.vbz_beratung.get_user_kontaktperson",
                'args': {},
                'async': false,
                'callback': function(r)
                {
                    frappe.route_options = {'status': 'Rückfragen', 'kontaktperson': ['in', r.message], 'ungelesen': 0, "sektion_id": sektion}
                    frappe.set_route("List", "Beratung", "List");
                }
            });
        });
        $("#p3").click(function(){
            frappe.call({
                'method': "mvd.mvd.page.vbz_beratung.vbz_beratung.get_user_kontaktperson",
                'args': {},
                'async': false,
                'callback': function(r)
                {
                    frappe.route_options = {'status': 'Rückfragen', 'kontaktperson': ['in', r.message], 'ungelesen': 1, "sektion_id": sektion}
                    frappe.set_route("List", "Beratung", "List");
                }
            });
        });
        $("#p4").click(function(){
            frappe.call({
                'method': "mvd.mvd.page.vbz_beratung.vbz_beratung.get_user_kontaktperson",
                'args': {},
                'async': false,
                'callback': function(r)
                {
                    frappe.route_options = {'status': 'Termin vereinbart', 'kontaktperson': ['in', r.message], 'hat_termine': 1, "sektion_id": sektion}
                    frappe.set_route("List", "Beratung", "List");
                }
            });
        });
        
        
        $("#rechtsberaterinnen").click(function(){
            frappe.set_route("List", "Termin Kontaktperson", "List");
        });
        $("#beratungskategorien").click(function(){
            frappe.set_route("List", "Beratungskategorie", "List");
        });
        $("#statistik").click(function(){
            frappe.set_route(["query-report", "Beratungsstatistik"]);
        });
        
        frappe.dom.unfreeze();
    }
}
